File: modyn/storage/internal/file_wrapper/csv_file_wrapper.py
import csv
from typing import Iterator, Optional

from modyn.storage.internal.file_wrapper.abstract_file_wrapper import AbstractFileWrapper
from modyn.storage.internal.file_wrapper.file_wrapper_type import FileWrapperType
from modyn.storage.internal.filesystem_wrapper.abstract_filesystem_wrapper import AbstractFileSystemWrapper


class CsvFileWrapper(AbstractFileWrapper):
    def __init__(self, file_path: str, file_wrapper_config: dict, filesystem_wrapper: AbstractFileSystemWrapper):
        super().__init__(file_path, file_wrapper_config, filesystem_wrapper)

        self.file_wrapper_type = FileWrapperType.CsvFileWrapper

        if "separator" in file_wrapper_config:
            self.separator = file_wrapper_config["separator"]
        else:
            self.separator = ","

        if "label_index" not in file_wrapper_config:
            raise ValueError("Please specify the index of the column that contains the label. ")
        if not isinstance(file_wrapper_config["label_index"], int) or file_wrapper_config["label_index"] < 0:
            raise ValueError("The label_index must be a positive integer.")
        self.label_index = file_wrapper_config["label_index"]

        # the first line might contain the header, which is useless and must not be returned.
        if "ignore_first_line" in file_wrapper_config:
            self.ignore_first_line = file_wrapper_config["ignore_first_line"]
        else:
            self.ignore_first_line = False

        if "encoding" in file_wrapper_config:
            self.encoding = file_wrapper_config["encoding"]
        else:
            self.encoding = "utf-8"

        # check that the file is actually a CSV
        self._validate_file_extension()

        # do not validate the content only if "validate_file_content" is explicitly set to False
        if ("validate_file_content" not in file_wrapper_config) or (
            "validate_file_content" in file_wrapper_config and file_wrapper_config["validate_file_content"]
        ):
            self._validate_file_content()

    def _validate_file_extension(self) -> None:
        """Validates the file extension as csv

        Raises:
            ValueError: File has wrong file extension
        """
        if not self.file_path.endswith(".csv"):
            raise ValueError("File has wrong file extension.")

    def _validate_file_content(self) -> None:
        """
        Performs the following checks:
        - specified label column is castable to integer
        - each row has the label_index_column
        - each row has the same width

        Raises a ValueError if a condition is not met
        """

        reader = self._get_csv_reader()

        number_of_columns = []

        for row in reader:
            number_of_columns.append(len(row))
            if not 0 <= self.label_index < len(row):
                raise ValueError("Label index outside row boundary")
            if not row[self.label_index].isnumeric():  # returns true iff all the characters are numbers
                raise ValueError("The label must be an integer")

        if len(set(number_of_columns)) != 1:
            raise ValueError(
                "Some rows have different width. " f"This is the number of columns row by row {number_of_columns}"
            )

    def get_sample(self, index: int) -> bytes:
        samples = self._filter_rows_samples([index])

        if len(samples) != 1:
            raise IndexError("Invalid index")

        return samples[0]

    def get_samples(self, start: int, end: int) -> list[bytes]:
        indices = list(range(start, end))
        return self.get_samples_from_indices(indices)

    def get_samples_from_indices(self, indices: list) -> list[bytes]:
        return self._filter_rows_samples(indices)

    def get_label(self, index: int) -> int:
        labels = self._filter_rows_labels([index])

        if len(labels) != 1:
            raise IndexError("Invalid index.")

        return labels[0]

    def get_all_labels(self) -> list[int]:
        reader = self._get_csv_reader()
        labels = [int(row[self.label_index]) for row in reader]
        return labels

    def get_number_of_samples(self) -> int:
        reader = self._get_csv_reader()
        return sum(1 for _ in reader)

    def _get_csv_reader(self) -> Iterator:
        """
        Receives the bytes from the file_system_wrapper and creates a csv.reader out of it.
        Returns:
            csv.reader
        """
        data_file = self.filesystem_wrapper.get(self.file_path)

        # Convert bytes content to a string
        data_file_str = data_file.decode(self.encoding)

        lines = data_file_str.split("\n")

        # Create a CSV reader
        reader = csv.reader(lines, delimiter=self.separator)

        # skip the header if required
        if self.ignore_first_line:
            next(reader)

        return reader

    def _filter_rows_samples(self, indices: list[int]) -> list[bytes]:
        """
        Filters the selected rows and removes the label column
        Args:
            indices: list of rows that must be kept

        Returns:
            list of byte-encoded rows

        """
        assert len(indices) == len(set(indices)), "An index is required more than once."
        reader = self._get_csv_reader()

        # Iterate over the rows and keep the selected ones
        filtered_rows: list[Optional[bytes]] = [None] * len(indices)
        for i, row in enumerate(reader):
            if i in indices:
                # Remove the label, convert the row to bytes and append to the list
                row_without_label = [col for j, col in enumerate(row) if j != self.label_index]
                # the row is transformed in a similar csv using the same separator and then transformed to bytes
                filtered_rows[indices.index(i)] = bytes(self.separator.join(row_without_label), self.encoding)

        if sum(1 for el in filtered_rows if el is None) != 0:
            raise IndexError("At least one index is invalid")

        # Here mypy complains that filtered_rows is a list of list[Optional[bytes]],
        # that can't happen given the above exception
        return filtered_rows  # type: ignore

    def _filter_rows_labels(self, indices: list[int]) -> list[int]:
        """
        Filters the selected rows and extracts the label column
        Args:
            indices: list of rows that must be kept

        Returns:
            list of labels

        """
        assert len(indices) == len(set(indices)), "An index is required more than once."
        reader = self._get_csv_reader()

        # Iterate over the rows and keep the selected ones
        filtered_rows: list[Optional[int]] = [None] * len(indices)
        for i, row in enumerate(reader):
            if i in indices:
                # labels are integer in modyn
                int_label = int(row[self.label_index])
                filtered_rows[indices.index(i)] = int_label

        if sum(1 for el in filtered_rows if el is None) != 0:
            raise IndexError("At least one index is invalid")

        # Here mypy complains that filtered_rows is a list of list[Optional[bytes]],
        # that can't happen given the above exception
        return filtered_rows  # type: ignore

    def delete_samples(self, indices: list) -> None:
        pass

File: modyn/storage/src/internal/file_wrapper/csv_file_wrapper.cpp
#include "internal/file_wrapper/csv_file_wrapper.hpp"

#include <rapidcsv.h>

#include <algorithm>
#include <numeric>
#include <stdexcept>

using namespace modyn::storage;

void CsvFileWrapper::setup_document(const std::string& path) {
  stream_ = filesystem_wrapper_->get_stream(path);
  auto sep_params = rapidcsv::SeparatorParams(separator_);
  sep_params.mQuoteChar = quote_;
  sep_params.mQuotedLinebreaks = allow_quoted_linebreaks_;
  doc_ = rapidcsv::Document(*stream_, label_params_, sep_params);
}

void CsvFileWrapper::validate_file_extension() {
  if (file_path_.substr(file_path_.find_last_of('.') + 1) != "csv") {
    FAIL("The file extension must be .csv");
  }
}

std::vector<unsigned char> CsvFileWrapper::get_sample(uint64_t index) {
  ASSERT(index < get_number_of_samples(), "Invalid index");

  std::vector<std::string> row = doc_.GetRow<std::string>(index);
  if (has_labels_) {
    row.erase(row.begin() + static_cast<int64_t>(label_index_));
  }
  if (has_targets_) {
    row.erase(row.begin() + static_cast<int64_t>(target_index_));
  }

  std::string row_string;
  for (const auto& cell : row) {
    row_string += cell + separator_;
  }
  row_string.pop_back();
  return {row_string.begin(), row_string.end()};
}

std::vector<std::vector<unsigned char>> CsvFileWrapper::get_samples(uint64_t start, uint64_t end) {
  ASSERT(end >= start && end <= get_number_of_samples(), "Invalid indices");

  std::vector<std::vector<unsigned char>> samples;
  for (uint64_t i = start; i < end; ++i) {
    std::vector<std::string> row = doc_.GetRow<std::string>(i);
    if (has_labels_) {
      row.erase(row.begin() + static_cast<int64_t>(label_index_));
    }
    if (has_targets_) {
      row.erase(row.begin() + static_cast<int64_t>(target_index_));
    }

    std::string row_string;
    for (const auto& cell : row) {
      row_string += cell + separator_;
    }
    row_string.pop_back();
    samples.emplace_back(row_string.begin(), row_string.end());
  }

  return samples;
}

std::vector<std::vector<unsigned char>> CsvFileWrapper::get_samples_from_indices(const std::vector<uint64_t>& indices) {
  std::vector<std::vector<unsigned char>> samples;
  for (const uint64_t index : indices) {
    std::vector<std::string> row = doc_.GetRow<std::string>(index);
    if (has_labels_) {
      row.erase(row.begin() + static_cast<int64_t>(label_index_));
    }
    if (has_targets_) {
      row.erase(row.begin() + static_cast<int64_t>(target_index_));
    }
    std::string row_string;
    for (const auto& cell : row) {
      row_string += cell + separator_;
    }
    row_string.pop_back();
    samples.emplace_back(row_string.begin(), row_string.end());
  }
  return samples;
}

int64_t CsvFileWrapper::get_label(uint64_t index) {
  ASSERT(index < get_number_of_samples(), "Invalid index");
  return doc_.GetCell<int64_t>(static_cast<size_t>(label_index_), static_cast<size_t>(index));
}

std::vector<int64_t> CsvFileWrapper::get_all_labels() {
  std::vector<int64_t> labels;
  if (!has_labels_) {
    return {};
  }
  const uint64_t num_samples = get_number_of_samples();
  for (uint64_t i = 0; i < num_samples; i++) {
    labels.push_back(get_label(i));
  }
  return labels;
}

std::vector<std::vector<unsigned char>> CsvFileWrapper::get_targets(uint64_t start, uint64_t end) {
  ASSERT(end >= start && end <= get_number_of_samples(), "Invalid indices");

  std::vector<std::vector<unsigned char>> targets;
  for (uint64_t i = start; i < end; ++i) {
    targets.push_back(get_target(i));
  }
  return targets;
}

std::vector<std::vector<unsigned char>> CsvFileWrapper::get_targets_from_indices(const std::vector<uint64_t>& indices) {
  std::vector<std::vector<unsigned char>> targets;
  for (const uint64_t index : indices) {
    targets.push_back(get_target(index));
  }
  return targets;
}

std::vector<unsigned char> CsvFileWrapper::get_target(uint64_t index) {
  ASSERT(index < get_number_of_samples(), "Invalid index");

  std::vector<std::string> row = doc_.GetRow<std::string>(index);
  if (target_index_ >= row.size()) {
    throw std::runtime_error("Target index out of range for this row");
  }
  std::string target_cell = row[target_index_];
  return {target_cell.begin(), target_cell.end()};
}

uint64_t CsvFileWrapper::get_number_of_samples() { return static_cast<uint64_t>(doc_.GetRowCount()); }

void CsvFileWrapper::delete_samples(const std::vector<uint64_t>& indices) {
  ASSERT(std::all_of(indices.begin(), indices.end(), [&](uint64_t index) { return index < get_number_of_samples(); }),
         "Invalid indices");

  std::vector<uint64_t> indices_copy = indices;
  std::sort(indices_copy.begin(), indices_copy.end(), std::greater<>());

  for (const size_t index : indices_copy) {
    doc_.RemoveRow(index);
  }

  doc_.Save(file_path_);
}

void CsvFileWrapper::set_file_path(const std::string& path) {
  file_path_ = path;

  if (stream_->is_open()) {
    stream_->close();
  }

  setup_document(path);
}

FileWrapperType CsvFileWrapper::get_type() { return FileWrapperType::CSV; }
